### PR TITLE
Add file locks in a few more locations

### PR DESCRIPTION
**Pull request by alexcrichton**
One was discovered by making `target_dir` return a `Filesystem`, the other was discovered by failing tests on the nightly OSX bots.

### CODE REVIEW

File: src/cargo/ops/cargo_clean.rs [hunk 1]
```diff
@@ -21,17 +21,24 @@ pub fn clean(manifest_path: &Path, opts: &CleanOptions) -> CargoResult<()> {
 
     // If we have a spec, then we need to delete some packages, otherwise, just
     // remove the whole target directory and be done with it!
+    //
+    // Note that we don't bother grabbing a lock here as we're just going to
+    // blow it all away anyway.
```
Comment (brson) [inline src/cargo/ops/cargo_clean.rs]: That's still a race. Doesn't this mean that if you run `cargo clean` while you are building that they'll clobber each other? Is there a technical reason it's hard to lock this?

Comment (alexcrichton) [inline src/cargo/ops/cargo_clean.rs]: Yeah this'll mean that they produce weird results most likely. This is somewhat difficult to lock as the target directory is somewhat granularly locked as opposed to "all at once". Figuring out all the locks we'd need to grab to lock the whole thing seemed more trouble than it was worth.

I also felt that in general we should do as much as we can to help reasonable concurrent requests of Cargo work out, but conflicting things like this seemed fine to me to cause some weird results.

**Comment by brson** [comment]
r=me

**Pull request merged**